### ISSUE TITLE
Implement forum searching

### ISSUE THREAD

**Issue by lizthegrey** [bravecollective/forums#53]
Full-text index most likely unless there's a fancy search indexing solution based on mongo that isn't the native thing.

**Comment by jboning** [comment]
Potential solutions:
http://sphinxsearch.com/about/sphinx/
http://openfts.sourceforge.net/
https://lucene.apache.org/solr/

I don't have any experience with any of these; if anyone can make any recommendations, that would be great. Otherwise I guess I'll have to do some research.

**Comment by ghost** [comment]
I'm running Sphinx in Production right now. No idea how to set it up, but it's pretty fast considering the massive dataset I have.

**Comment by amcgregor** [comment]
Here, [have a solution almost designed for use with MongoEngine](https://gist.github.com/amcgregor/722c1075ba424092c799) ('cause it was).  It's a full Okapi BM-25 parallel ranking algorithm with binary search optimization (basically identical to Yahoo's and Lucene's) and sample generic (non-model-speicfic with minor adjustments) full text, weighted indexer.  (That does all the stemming, stop words, and text cleanup.)

Edit: Scratch that, it is generic.  ObjectIds won't collide, but you'll need to add some way of identifying the source.  (Use of full ReferenceFields is not recommended unless changes are made to the ranking algorithm; you don't want all of those documents silently loaded during ranking like that!)

**Comment by jboning** [comment]
To loop around on this for anyone who wasn't watching IRC, I plugged in @amcgregor's solution and got it working, but came to the conclusion that we don't want to be maintaning/extending a homegrown search implementation. I'd rather pull in an external, fully-featured solution (with support for e.g. faceted search, hit highlighting, natural queries). It'll be much heavier-weight and require more setup, but I think it's the right tradeoff here. After some scouting of the internet, I decided to go with Solr and #108 is a PR covering the backend communication with Solr.
